Fix System.DirectoryServices.Protocols to work on Linux

psteniusubi opened the pull request:
Fix #66243 

LdapPal.CompareDirectoryEntries on Linux does not pass strValue to Interop.Ldap.ldap_compare

This fix always converts CompareRequest.Assertion to byte array

psteniusubi committed:
Always convert CompareRequest.Assertion to byte array when invoking LdapPal.CompareDirectoryEntries

LdapPal.CompareDirectoryEntries on Linux does not pass strValue to Interop.Ldap.ldap_compare

psteniusubi committed:
Merge branch 'main' into issue-66243

psteniusubi committed:
add integration test for CompareRequest

psteniusubi committed:
Merge branch 'issue-66243' of github.com:psteniusubi/runtime into issue-66243

psteniusubi committed:
fix formatting

psteniusubi committed:
cross platform fixes to integration tests

psteniusubi committed:
add more integration tests

psteniusubi committed:
Merge branch 'dotnet:main' into issue-66243

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67338) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67338) before we can accept your contribution.<br/><br/>:x: psteniusubi [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=67338)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=67338) it.</sub>

gfoidl commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
+                    }
+                    else if (assertion[0] is Uri uri)
+                    {
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(uri.ToString());
```
```suggestion
                        byteArray = Encoding.UTF8.GetBytes(uri.ToString());
```

gfoidl commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
+                        bv_len = byteArray.Length,
+                        bv_val = Marshal.AllocHGlobal(byteArray.Length)
+                    };
+                    Marshal.Copy(byteArray, 0, berValuePtr.bv_val, byteArray.Length);
```
I wonder if `byte[] byteArray` is needed or if this can be stackallocated (and the copied)?

psteniusubi replied:
This code follows how BuildAttributes method does the same conversions. I don't think improving this is within scope of this fix.

gfoidl replied:
> improving this is within scope of this fix.

You are right, that's going beyond the fix.

The code here seems pretty old, and was written before span came around. So there could be a general overhaul -- but I don't know the plans here and if it is worth, as such an upgrade comes with cost and this code seems to be battle-tested, which a re-write has to prove first.

joperezr replied:
I don't think we should stackalloc this particular array since it is based on user input, so we want to make sure that we don't call stackalloc from arbitriary input size.

psteniusubi replied:
Using something like Marshal.StringToCoTaskMemUTF8 could save one allocation. But as I said I don't think this should be within scope of this fix.

psteniusubi commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
-                            };
-                            Marshal.Copy(byteArray, 0, berValuePtr.bv_val, byteArray.Length);
-                        }
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(str);
```
Hi @gfoidl. Thank you for your comments. I'm following style and conventions elsewhere in this component. More specifically implementation of BuildAttributes method:

https://github.com/dotnet/runtime/blob/d64b8af65e7eecc52bf81d40315fa80c64c818c1/src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs#L1277

joperezr commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
+                    }
+                    else if (assertion[0] is Uri uri)
+                    {
+                        var encoder = new UTF8Encoding();
```
Looks like the test you are adding is only covering when the assertion is a string. Can you please turn your Fact into a Theory instead that uses both string and URI in order to make sure we have coverage for both?

joperezr commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
                     else
                     {
-                        stringValue = assertion[0].ToString();
+                        throw new ArgumentException(SR.ValidValueType);
```
Can we also add a tests that ensures that this Argument Exception gets thrown?

psteniusubi replied:
I had a look at this. Turns out DirectoryAttribute rejects other value types than string, Uri and byte[]. So this condition never happens.

joperezr commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs:
```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
+                    }
+                    else if (assertion[0] is byte[] bytes)
+                    {
+                        byteArray = bytes;
```
What happens if the bytes were not utf8 encoded?

psteniusubi replied:
CompareRequest operation returns ResultCode.CompareFalse. I added a test case for this.

joperezr replied:
Right, but I guess my question is: should we support that? Or is it expected that bytes will always be UTF8 encoded? I quickly looked at our [docs](https://docs.microsoft.com/en-us/dotnet/api/system.directoryservices.protocols.comparerequest.-ctor?view=dotnet-plat-ext-6.0#system-directoryservices-protocols-comparerequest-ctor(system-string-system-string-system-byte())) and we don't seem to mention anywhere that the bytes need to be UTF8 encoded.

joperezr replied:
I don't want to block your contribution on that, if we think UTF8 is what should be expected then we can just update our docs to point that out, and perhaps consider in the future add another constructor to CompareRequest that takes in bytes[] and Encoding.

psteniusubi replied:
String values are always utf-8 encoded on the wire (per [ldap spec](https://datatracker.ietf.org/doc/html/rfc4511#section-4.1.2)). The byte[] value is as-is on the wire.

joperezr replied:
Sounds good, we can update the docs then to call this out, but we can do that separately from your PR

psteniusubi committed:
implemented TestCompareRequestTheory as ConditionalTheory

test data from TestCompareRequestTheory_TestData
add cases with CompreRequest Uri value

joperezr reviewed (approved):
Thanks for the contribution!

danmoseley commented:
Yes, thanks @psteniusubi . We would welcome other contributions, if you are interested.
